src/lib.rs

```diff
@@ -82,6 +89,54 @@ impl Deref for AllocLock {
     }
 }
 
+unsafe impl GlobalAlloc for GCMalloc {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        let ptr = libc::malloc(layout.size() as libc::size_t) as *mut u8;
+
+        ALLOC_LOCK.lock();
+        match ALLOC_INFO {
+            Some(ref mut pm) => pm.insert(ptr as usize, layout.size()),
+            None => {
+                let mut al = AllocList::new();
+                al.insert(ptr as usize, layout.size());
+                ALLOC_INFO = Some(al);
+            }
+        };
+        ALLOC_LOCK.unlock();
+        ptr
+    }
+
+    unsafe fn dealloc(&self, ptr: *mut u8, _layout: Layout) {
+        libc::free(ptr as *mut libc::c_void);
```

Comment:
Similar question here about atomicity.

Reply:
Again, I don't think `dealloc` needs to happen atomically. If the thread is preempted before acquiring the lock, it will just mean that the ptr info is removed from `ALLOC_INFO` after another allocation has finished.

Reply:
I think we need to define precisely when code can access the list of memory stuff and what it can do with it. For example, we can define that it's safe for code to add and remove stuff to `ALLOC_INFO` at will (provided `ALLOC_LOCK` is held) if the only time we actually make use of the stuff in `ALLOC_INFO` is during a stop-the-world GC (which, by definition, is an implicit lock).

However, in contrast, we might allow threads to read from `ALLOC_INFO` outside of a stop-the-world GC at which point a thread might be interrupted after `free` but before `ALLOC_LOCK.lock()`, leading to a possible read-after-free possibility. If that's the case then we would need to do something like:

```rust
ALLOC_LOCK.lock()
free(...)
ALLOC_INFO....remove(...);
ALLOC_LOCK.unlock()
```

I don't have a firm opinion on what the best thing to do here is, but I'm fairly sure we need to explicitly pick one choice.

Reply:
Right, Laurie and I essentially have the same query:  "Should the malloc/free's be inside the critical section?"

Reply:
Right now, I can only ever see the need to read from this table during a STW pause, which, as you've mentioned, implicitly locks `ALLOC_INFO`. I think making this assumption is okay for now providing it's well documented on any read method.

Reply:
OK.

Reply:
Let's put an explicit comment about STW somewhere then, so we're clear in this PR, and clear going forward.

Reply:
Done